### ISSUE TITLE
Question on the for loop in forward pass

### ISSUE THREAD

**Issue by Chuhanxx** [jayleicn/ClipBERT#9]
Hi, Jie. Thank you for doing this excellent work and publishing the code. 
I have one question regarding the fine-tuning for downstream tasks. I noticed that N clips sampled from each video are forwarded individually using a for loop: https://github.com/jayleicn/ClipBERT/blob/main/src/tasks/run_video_qa.py#L250

May I ask what is the purpose of forwarding these clips separately instead of grouping them as batch_size * n_clips?

**Comment by jayleicn** [comment]
Hi Chuhan,

Whether grouping them or not does not affect the final results. We write it this way for (1) ease of implementation; and to (2) reduce memory cost. Grouping them as `batch_size * n_clips` will require adjusting`batch_size` for different `n_clips`, which is a bit annoying, while using a for-loop, we can simply tune to the max `batch_size` for a given GPU and keep it the same for different `n_clips`.

Hope this helps!
Jie

**Comment by Chuhanxx** [comment]
Thank you for the reply, it answers my question. 

Just an additional note: from my shallow understanding, it does not save a lot of memory if only summing up the loss without accumulating the gradients (i.e, putting loss.backward outside the loop). 
Reference: 
https://discuss.pytorch.org/t/why-do-we-need-to-set-the-gradients-manually-to-zero-in-pytorch/4903/20
https://stackoverflow.com/questions/62067400/understanding-accumulated-gradients-in-pytorch
But it is not a big issue if memory cost and training time is not a bottleneck in the experiments. 

Thanks again and good luck!

**Comment by jayleicn** [comment]
Thanks for your references. Yes, you are right, it does not save much memory for training. I mistakenly only answered your question in the case of inference, not training, since this line https://github.com/jayleicn/ClipBERT/blob/main/src/tasks/run_video_qa.py#L250 points to the validate loop. 

At training time, it mostly only eases our implementation.
